fe/fe-core/src/main/java/com/starrocks/alter/dynamictablet/SplitTabletJobFactory.java

```diff
@@ -0,0 +1,314 @@
+import com.starrocks.catalog.PhysicalPartition;
+import com.starrocks.catalog.Tablet;
+import com.starrocks.catalog.TabletInvertedIndex;
+import com.starrocks.catalog.TabletMeta;
+import com.starrocks.common.Config;
+import com.starrocks.common.StarRocksException;
+import com.starrocks.common.util.concurrent.lock.LockType;
+import com.starrocks.common.util.concurrent.lock.Locker;
+import com.starrocks.lake.LakeTablet;
+import com.starrocks.server.GlobalStateMgr;
+import com.starrocks.server.WarehouseManager;
+import com.starrocks.sql.ast.SplitTabletClause;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+/*
+ * SplitTabletJobFactory is for creating SplitTabletJob.
+ */
+public class SplitTabletJobFactory implements DynamicTabletJobFactory {
+    private static final Logger LOG = LogManager.getLogger(SplitTabletJobFactory.class);
+
+    private final Database db;
+
+    private final OlapTable table;
+
+    private final SplitTabletClause splitTabletClause;
+
+    public SplitTabletJobFactory(Database db, OlapTable table, SplitTabletClause splitTabletClause) {
+        this.db = db;
+        this.table = table;
+        this.splitTabletClause = splitTabletClause;
+    }
+
+    /*
+     * Create a dynamic tablet job and return it.
+     * Dynamic tablets are created for all related materialized indexes.
+     * New shareds are created for new tablets.
+     */
+    @Override
+    public DynamicTabletJob createDynamicTabletJob() throws StarRocksException {
+        if (!table.isCloudNativeTableOrMaterializedView()) {
+            throw new StarRocksException("Unsupported table type " + table.getType()
+                    + " in table " + db.getFullName() + '.' + table.getName());
+        }
+
+        Map<PhysicalPartition, Collection<MaterializedIndex>> materializedIndexes = createDynamicTablets();
+        if (materializedIndexes.isEmpty()) {
+            throw new StarRocksException("No tablets need to split in table "
+                    + db.getFullName() + '.' + table.getName());
+        }
+
+        try {
+            createNewShards(materializedIndexes);
+            Map<Long, DynamicTabletContext> dynamicTabletContexts = createDynamicTabletContexts(materializedIndexes);
+            long jobId = GlobalStateMgr.getCurrentState().getNextId();
+            return new SplitTabletJob(jobId, db.getId(), table.getId(), dynamicTabletContexts);
+        } catch (Exception e) {
+            clearDynamicTablets(materializedIndexes);
+            throw e;
+        }
+    }
+
+    /*
+     * Create dynamic tablets for all related materialized indexes
+     * Return all the related materialized indexes
+     */
+    private Map<PhysicalPartition, Collection<MaterializedIndex>> createDynamicTablets() throws StarRocksException {
```

Comment:
You return a large number of metadata objects as parameters, and these objects will escape the protection of meta lock, which will bring great risks to future development。Can it be done more elegantly?